# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/notification/NotificationParent.cpp
# Commit: fca3921d09a5
# Full Hash: fca3921d09a59b709a6939f4a1e65f261367e0a6
# Author: Kagami Sascha Rosylight <saschanaz@outlook.com>
# Date: 2024-11-12 09:32:50
# Regressor Bug: 1928702
# File Overlap Count: 1
# Description:
#   Bug 1928702 - Part 5: Implement RecvShow r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D227690
# ==============================================================================

diff -r 807f1d1ada66 -r fca3921d09a5 dom/notification/NotificationParent.cpp
--- a/dom/notification/NotificationParent.cpp	Mon Nov 11 20:34:12 2024 +0000
+++ b/dom/notification/NotificationParent.cpp	Mon Nov 11 20:34:13 2024 +0000
@@ -8,6 +8,9 @@
 
 #include "NotificationUtils.h"
 #include "mozilla/ipc/Endpoint.h"
+#include "mozilla/Components.h"
+#include "nsComponentManagerUtils.h"
+#include "nsIAlertsService.h"
 
 namespace mozilla::dom::notification {
 
@@ -19,10 +22,69 @@
   return NS_OK;
 }
 
+// Step 4 of
+// https://notifications.spec.whatwg.org/#dom-notification-notification
 mozilla::ipc::IPCResult NotificationParent::RecvShow(ShowResolver&& aResolver) {
+  // Step 4.2: Run the fetch steps for notification. (Will happen in
+  // nsIAlertNotification::LoadImage)
+  // Step 4.3: Run the show steps for notification.
+  Show();
   return IPC_OK();
 }
 
+nsresult NotificationParent::Show() {
+  // Step 4.3 the show steps, which are almost all about processing `tag` and
+  // then displaying the notification. Both are handled by
+  // nsIAlertsService::ShowAlert/PersistentNotification. The below is all about
+  // constructing the observer (for show and close events) right and ultimately
+  // call the alerts service function.
+
+  // XXX(krosylight): Non-persistent notifications probably don't need this
+  nsAutoString alertName;
+  GetAlertName(alertName);
+  nsresult rv =
+      PersistNotification(mPrincipal, mId, alertName, mOptions, mScope);
+  if (NS_FAILED(rv)) {
+    NS_WARNING("Could not persist Notification");
+  }
+
+  // In the case of IPC, the parent process uses the cookie to map to
+  // nsIObserver. Thus the cookie must be unique to differentiate observers.
+  // XXX(krosylight): This is about ContentChild::mAlertObserver which is not
+  // useful when called by the parent process. This should be removed when we
+  // make nsIAlertsService parent process only.
+  nsString obsoleteCookie = u"notification:"_ns;
+
+  bool requireInteraction = mOptions.requireInteraction();
+  if (!StaticPrefs::dom_webnotifications_requireinteraction_enabled()) {
+    requireInteraction = false;
+  }
+
+  nsCOMPtr<nsIAlertNotification> alert =
+      do_CreateInstance(ALERT_NOTIFICATION_CONTRACTID);
+  if (!alert) {
+    return NS_ERROR_NOT_AVAILABLE;
+  }
+  MOZ_TRY(alert->Init(alertName, mOptions.icon(), mOptions.title(),
+                      mOptions.body(), true, obsoleteCookie,
+                      NS_ConvertASCIItoUTF16(GetEnumString(mOptions.dir())),
+                      mOptions.lang(), mOptions.dataSerialized(), mPrincipal,
+                      mPrincipal->GetIsInPrivateBrowsing(), requireInteraction,
+                      mOptions.silent(), mOptions.vibrate()));
+
+  nsCOMPtr<nsIAlertsService> alertService = components::Alerts::Service();
+
+  // XXX: We previously used ShowPersistentNotification here, but on all desktop
+  // backends and PContent it has always been just a pure alias to ShowAlert.
+  // And using it properly causes incorrect behavior on Android listener
+  // registration. Given the persistence happens elsewhere via
+  // UnpersistNotification(), we should remove
+  // nsIAlertsService::ShowPersistentNotification.
+  MOZ_TRY(alertService->ShowAlert(alert, this));
+
+  return NS_OK;
+}
+
 nsresult NotificationParent::BindToMainThread(
     Endpoint<PNotificationParent>&& aParentEndpoint,
     PBackgroundParent::CreateNotificationParentResolver&& aResolver) {